src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,376 @@
+    """Convert SRE regex parse tree to strategy that generates strings matching
+    that regex represented by that parse tree.
+
+    `codes` is either a list of SRE regex elements representations or a
+    particular element representation. Each element is a tuple of element code
+    (as string) and parameters. E.g. regex 'ab[0-9]+' compiles to following
+    elements:
+
+        [
+            ('literal', 97),
+            ('literal', 98),
+            ('max_repeat', (1, 4294967295, [
+                ('in', [
+                    ('range', (48, 57))
+                ])
+            ]))
+        ]
+
+    The function recursively traverses regex element tree and converts each
+    element to strategy that generates strings that match that element.
+
+    Context stores
+    1. List of groups (for backreferences)
+    2. Active regex flags (e.g. IGNORECASE, DOTALL, UNICODE, they affect
+       behavior of various inner strategies)
+
+    """
+    def recurse(codes):
+        return _strategy(codes, context, is_text)
+
+    if is_text:
+        EMPTY = u''
+        to_char = hunichr
+    else:
+        EMPTY = b''
+
+        def to_char(i):
+            return bytes([i])
+
+    if not isinstance(codes, tuple):
+        # List of codes
+        strategies = []
+
+        i = 0
+        while i < len(codes):
+            if codes[i][0] == sre.LITERAL and \
+                    not context.flags & re.IGNORECASE:
+                # Merge subsequent "literals" into one `just()` strategy
+                # that generates corresponding text if no IGNORECASE
+                j = i + 1
+                while j < len(codes) and codes[j][0] == sre.LITERAL:
+                    j += 1
+
+                if i + 1 < j:
+                    strategies.append(hs.just(
+                        EMPTY.join([to_char(charcode)
+                                    for (_, charcode) in codes[i:j]])
+                    ))
+
+                    i = j
+                    continue
+
+            strategies.append(recurse(codes[i]))
+            i += 1
+
+        if not strategies:
+            return hs.just(EMPTY)
+        if len(strategies) == 1:
+            return strategies[0]
+        return hs.tuples(*strategies).map(EMPTY.join)
+    else:
+        # Single code
+        code, value = codes
+        if code == sre.LITERAL:
+            # Regex 'a' (single char)
+            c = to_char(value)
+            if context.flags & re.IGNORECASE:
+                return hs.sampled_from([c.lower(), c.upper()])
+            return hs.just(c)
+
+        elif code == sre.NOT_LITERAL:
+            # Regex '[^a]' (negation of a single char)
+            c = to_char(value)
+            blacklist = set([c.lower(), c.upper()]) \
+                if context.flags & re.IGNORECASE else [c]
+            if is_text:
+                return hs.characters(blacklist_characters=blacklist)
+            else:
+                return hs.binary().filter(lambda c: c not in blacklist)
```

Comment:
I'm confused as to why this returns only single element strings for text but strings of any length for binary. It looks like it's probably a bug, but I don't think I understand this code well enough to say for sure.